### PR TITLE
Revisit caching for RazorViewEngine

### LINKED ISSUE #3337: Revisit caching in RazorViewEngine
The primary motivation of this work item is perf, but there's also a desire to remedy some of the experience gaps in using Razor views when files are added or deleted on disk.

#### Experience gaps
The Razor view engine caches positive and negative path lookups produced by the ViewLocationExpander. For positive results we do not look at any paths besides the successful one unless the file at path no longer exists. For e.g.

/Views/Home/Index.cshtml -> Does not exist
/Views/Shared/Index.cshtml -> Exists

We'll cache the path /Views/Shared/Index.cshtml. Unless /Views/Shared/Index.cshtml is deleted, we do not look at /Views/Home/Index.cshtml. This ends up being poor dev experience when files get moved around.

#### Caching _ViewStarts
As part of executing a (not partial) RazorView, we always reconstruct all potential ViewStart file paths. These need to be cached since they show up in allocation profiles.

#### Design change
Based on a discussion with @loudej and @rynowak, here's the proposal for how we'd like the this to behave:

* Remove IViewStartProvider and make it a behavior of RazorViewEngine. (`protected RazorViewEngine.GetViewStartPages`)

* We'll cache view start locations, the result of IRazorPageFactory.CreateInstance, and the IRazorPage[] instances for _ViewStarts that are required to execute the RazorView. 

* The expiration trigger for these cached entries would include the result of calling IFileProvider.Watch on all paths that were looked up trying to look up the view, _ViewImports that were used to generate the view and all potential _ViewStart locations.

* The cache will expire if a file is added at a path with higher precedence and solves the experience gap. It'll additionally cache _ViewStarts removing the need to look these up on a per request basis.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #3337

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @rynowak

**Comment by rynowak** [comment]
Sadly I won't get to this until later tonite.

**Comment by pranavkm** [comment]
:up::date:

**Comment by rynowak** [comment]
9 days later, I'm getting to this :+1:

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs [hunk 1]
```diff
@@ -54,9 +55,10 @@ public CompilerCache(IFileProvider fileProvider)
                 throw new ArgumentNullException(nameof(precompiledViews));
             }
 
+            var expirationTokens = new IChangeToken[0];
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs]: Might be nice to add a comment calling out that we're explicitly calling out precompiled views as non-expiring

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCacheResult.cs [hunk 2]
```diff
@@ -11,31 +13,35 @@ namespace Microsoft.AspNet.Mvc.Razor.Compilation
...
-        /// </summary>
-        public static CompilerCacheResult FileNotFound { get; } = new CompilerCacheResult();
-
-        /// <summary>
         /// Initializes a new instance of <see cref="CompilerCacheResult"/> with the specified
         /// <see cref="CompilationResult"/>.
         /// </summary>
         /// <param name="compilationResult">The <see cref="Compilation.CompilationResult"/> </param>
-        public CompilerCacheResult(CompilationResult compilationResult)
+        public CompilerCacheResult(CompilationResult compilationResult, IList<IChangeToken> expirationTokens)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCacheResult.cs]: nit: Should we also just add a constructor for the no-token case (precompilation)?

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCacheResult.cs [hunk 3]
```diff
@@ -43,5 +49,9 @@ protected CompilerCacheResult()
         /// </summary>
         /// <remarks>This property is null when file lookup failed.</remarks>
         public CompilationResult CompilationResult { get; }
+
+        public IList<IChangeToken> ExpirationTokens { get; }
+
+        public bool IsFoundResult => CompilationResult != null;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCacheResult.cs]: `WasViewFound`?

File: src/Microsoft.AspNet.Mvc.Razor/DefaultRazorPageFactory.cs [hunk 4]
```diff
@@ -10,7 +10,7 @@ namespace Microsoft.AspNet.Mvc.Razor
     /// Represents a <see cref="IRazorPageFactory"/> that creates <see cref="RazorPage"/> instances
     /// from razor files in the file system.
     /// </summary>
-    public class VirtualPathRazorPageFactory : IRazorPageFactory
+    public class DefaultRazorPageFactory : IRazorPageFactory
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultRazorPageFactory.cs]: :dancer:

File: src/Microsoft.AspNet.Mvc.Razor/DefaultRazorPageFactory.cs [hunk 5]
```diff
@@ -58,18 +58,20 @@ public IRazorPage CreateInstance(string relativePath)
...
-            if (result == CompilerCacheResult.FileNotFound)
+            if (result.IsFoundResult)
             {
-                return null;
+                return new RazorPageFactoryResult(() =>
+                {
+                    var page = (IRazorPage)Activator.CreateInstance(result.CompilationResult.CompiledType);
+                    page.Path = relativePath;
+                    return page;
+                }, result.ExpirationTokens);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultRazorPageFactory.cs]: What's up with this wonky lambda thing? Why doesn't this class just take the required parameters in the constructor?

File: src/Microsoft.AspNet.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs [hunk 6]
```diff
@@ -148,9 +146,8 @@ internal static void AddRazorViewEngineServices(IServiceCollection services)
             // In the default scenario the following services are singleton by virtue of being initialized as part of
             // creating the singleton RazorViewEngine instance.
             services.TryAddTransient<IRazorViewFactory, RazorViewFactory>();
-            services.TryAddTransient<IRazorPageFactory, VirtualPathRazorPageFactory>();
+            services.TryAddTransient<IRazorPageFactory, DefaultRazorPageFactory>();
             services.TryAddTransient<IRazorCompilationService, RazorCompilationService>();
-            services.TryAddTransient<IViewStartProvider, ViewStartProvider>();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/DependencyInjection/MvcRazorMvcCoreBuilderExtensions.cs]: :dancer:

File: src/Microsoft.AspNet.Mvc.Razor/IRazorPageFactory.cs [hunk 7]
```diff
@@ -9,10 +9,10 @@ namespace Microsoft.AspNet.Mvc.Razor
...
     {
         /// <summary>
-        /// Creates a <see cref="IRazorPage"/> for the specified path.
+        /// Creates a <see cref="IRazorPage"/> factory for the specified path.
         /// </summary>
         /// <param name="relativePath">The path to locate the page.</param>
-        /// <returns>The IRazorPage instance if it exists, null otherwise.</returns>
-        IRazorPage CreateInstance(string relativePath);
+        /// <returns>The <see cref="RazorPageFactoryResult"/> instance.</returns>
+        RazorPageFactoryResult CreateFactory(string relativePath);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/IRazorPageFactory.cs]: Doesn't that make this a factory factory?

File: src/Microsoft.AspNet.Mvc.Razor/DefaultRazorPageFactory.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultRazorPageFactory.cs]: Or make this a protected method and pass in a handle to it.

File: src/Microsoft.AspNet.Mvc.Razor/IRazorViewFactory.cs [hunk 8]
```diff
@@ -18,6 +19,6 @@ public interface IRazorViewFactory
         /// <param name="page">The <see cref="IRazorPage"/> instance to execute.</param>
         /// <param name="isPartial">Determines if the view is to be executed as a partial.</param>
         /// <returns>A <see cref="IView"/> instance that renders the contents of the <paramref name="page"/></returns>
-        IView GetView(IRazorViewEngine viewEngine, IRazorPage page, bool isPartial);
+        IView GetView(IRazorViewEngine viewEngine, IRazorPage page, IReadOnlyList<IRazorPage> viewStarts, bool isPartial);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/IRazorViewFactory.cs]: This doesn't seem useful as an extensibility point, can we just move this code into the view engine directly?

File: src/Microsoft.AspNet.Mvc.Razor/RazorPageFactoryResult.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        {
+            RazorPageFactory = razorPageFactory;
+            ExpirationTokens = expirationTokens;
+        }
+
+        public Func<IRazorPage> RazorPageFactory { get; }
+
+        public IList<IChangeToken> ExpirationTokens { get; }
+
+        public bool IsFoundResult => RazorPageFactory != null;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPageFactoryResult.cs]: This name is kinda gross too

File: src/Microsoft.AspNet.Mvc.Razor/RazorPageResult.cs [hunk 10]
```diff
@@ -9,7 +9,7 @@ namespace Microsoft.AspNet.Mvc.Razor
     /// <summary>
     /// Represents the results of locating a <see cref="IRazorPage"/>.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPageResult.cs]: Ugh. Represents.

File: src/Microsoft.AspNet.Mvc.Razor/RazorView.cs [hunk 11]
```diff
@@ -153,14 +153,13 @@ private Task RenderPageCoreAsync(IRazorPage page, ViewContext context)
...
         private async Task RenderViewStartAsync(ViewContext context)
         {
-            var viewStarts = _viewStartProvider.GetViewStartPages(RazorPage.Path);
-
             string layout = null;
             var oldFilePath = context.ExecutingFilePath;
             try
             {
-                foreach (var viewStart in viewStarts)
+                for (var i = 0; i < _viewStartPages.Count; i++)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorView.cs]: :clap:

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngine.cs [hunk 12]
```diff
@@ -72,10 +62,11 @@ public class RazorViewEngine : IRazorViewEngine
...
         /// <c>/Views/Home/Test.cshtml</c>. Locations such as <c>/views/home/test.cshtml</c> would not be discovered
         /// </remarks>
-        public virtual IEnumerable<string> ViewLocationFormats
+        public virtual IEnumerable<string> ViewLocationFormats { get; } = new[]
         {
-            get { return _viewLocationFormats; }
-        }
+            "/Views/{1}/{0}" + ViewExtension,
+            "/Views/Shared/{0}" + ViewExtension,
+        };
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngine.cs]: 18 PranavPoints :registered: awarded. Thank you for your dilligence

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngine.cs [hunk 13]
```diff
@@ -241,13 +240,32 @@ public virtual IEnumerable<string> AreaViewLocationFormats
                     applicationRelativePath += ViewExtension;
                 }
 
-                var page = _pageFactory.CreateInstance(applicationRelativePath);
-                if (page != null)
+                var cacheKey = new ViewLocationCacheKey(
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngine.cs]: This might be a good candidate for a factory method

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCacheResult.cs (hunk 3 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCacheResult.cs]: Success?

File: src/Microsoft.AspNet.Mvc.Razor/IRazorPageFactory.cs (hunk 7 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/IRazorPageFactory.cs]: Yeah. We could figure out if this needs to be renamed.

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/IRazorPageFactory.cs]: cc @Eilon, what do you think about `IRazorPageFactoryProvider`?

Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Razor/IRazorPageFactory.cs]: Sure, whatever, my head hurts just thinking about it :smile:

**Comment by pranavkm** [comment]
:up: :date:

**Comment by rynowak** [comment]
11 days later I'm finally getting to this :+1:

**Comment by rynowak** [comment]
:shipit:

**Pull request closed without merging**